lib/sycamore/sycamore/transforms/summarize.py

```diff
@@ -57,137 +58,375 @@ class LLMElementTextSummarizer(Summarizer):
-        document.properties["summary"] = summary
-        return document
-
-    def get_text(self, doc: Document) -> str:
-        doc_text = ""
-        props_dict = doc.properties.get("entity", {})
-        props_dict.update({p: doc.properties[p] for p in set(doc.properties) - set(BASE_PROPS)})
-        for k, v in props_dict.items():
-            doc_text += f"{k}: {v}\n"
-
-        doc_text_representation = ""
-        if not self.use_elements:
-            if doc.text_representation is not None:
-                doc_text_representation += doc.text_representation[:NUM_TEXT_CHARS_GENERATE]
-        else:
-            for element in doc.elements[: self.num_elements]:
-                # Greedy fill doc level text length
-                if len(doc_text_representation) >= NUM_TEXT_CHARS_GENERATE:
-                    break
-                doc_text_representation += (element.text_representation or "") + "\n"
-        doc_text += f"Text contents:\n{doc_text_representation}\n"
-
-        return doc_text
+        assert EtCetera not in fields[:-1], "EtCetera must be at the end of the list of fields if provided"
+        self.fields = fields
+        self.prompt = OneStepSummarizerPrompt.fork(**self.get_const_vars())
+
+    @staticmethod
+    def get_const_vars() -> dict[str, str]:
+        return {
+            "fields_key": "_fields",
+            "numel_key": "_num_elements",
+            "startel_key": "_start_element",
+        }
+
+    def preprocess(self, doc: Document) -> Document:
+        vars = self.get_const_vars()
+        prompt = self.prompt.fork(ignore_none=True, question=self.question)
+        fields = copy.deepcopy(self.fields)
+        etc = False
+        if len(fields) > 0 and fields[-1] is EtCetera:
+            etc = True
+            fields = fields[:-1]
+        all_element_property_names = {f"properties.{k}" for e in doc.elements for k in e.properties}
+        # Compute baseline 'fluff' tokens by setting fields and elements to 'no fields'
+        # and 'no elements'. Use this later to figure out how many tokens adding a field adds
+        doc.properties[vars["fields_key"]] = []
+        doc.properties[vars["numel_key"]] = 0
+        doc.properties[vars["startel_key"]] = 0
+        data_independent_ntk = prompt.render_document(doc).token_count(self.tokenizer)
+        # If fields is specified these are always included, so this will be our starting token total
+        doc.properties[vars["fields_key"]] = fields
+        curr_ntks = prompt.render_document(doc).token_count(self.tokenizer)
+        if etc:
+            for p in all_element_property_names:
+                if p in fields:
+                    continue
+                doc.properties[vars["fields_key"]] = [p]
+                ntk = prompt.render_document(doc).token_count(self.tokenizer) - data_independent_ntk
+                if curr_ntks + ntk < self.token_limit:
```

Comment:
Add a TODO to make this not O(n^2); I'm not sure how much it can hurt. The bad case is a lot of very small properties.

Either fix it to not be O(n^2), which I think will be a little messy since you need a "render only property and render only element controls"; or add a test case with 1000 documents, 26^2 2 character key -> 1 character value properties and a token limit of ~500k. Lets see how slow it is as to whether it's necessary to fix. If it can do that under 5s, the O(n^2) doesn't matter for now.

Reply:
I mis-read this. Yea, it's incremental. No need to write a test until we run into a problem for real.